zio-http/src/main/scala/zhttp/http/HttpData.scala

```diff
@@ -68,33 +62,90 @@ object HttpData {
+  }
 
   /**
    * Helper to create HttpData from Stream of string
    */
-  def fromStream(stream: ZStream[Any, Throwable, String], charset: Charset = HTTP_CHARSET): HttpData =
-    HttpData.BinaryStream(stream.map(str => Unpooled.copiedBuffer(str, charset)))
+  def fromStream(stream: ZStream[Any, Throwable, CharSequence], charset: Charset = HTTP_CHARSET): HttpData =
+    HttpData.BinaryStream(stream.map(str => Unpooled.wrappedBuffer(str.toString.getBytes(charset))))
 
   /**
-   * Helper to create HttpData from String
+   * Helper to create HttpData from Stream of bytes
    */
-  def fromString(text: String, charset: Charset = HTTP_CHARSET): HttpData = Text(text, charset)
+  def fromStream(stream: ZStream[Any, Throwable, Byte]): HttpData =
+    HttpData.BinaryStream(stream.mapChunks(chunks => Chunk(Unpooled.wrappedBuffer(chunks.toArray))))
 
   /**
-   * Helper to create HttpData from contents of a file
+   * Helper to create HttpData from String
    */
-  def fromFile(file: => java.io.File): HttpData = {
-    RandomAccessFile(() => new java.io.RandomAccessFile(file, "r"))
+  def fromString(text: String, charset: Charset = HTTP_CHARSET): HttpData = Text(text, charset)
+
+  private[zhttp] sealed trait Outgoing extends HttpData { self =>
+    def encode: ZIO[Any, Throwable, ByteBuf] =
+      self match {
+        case HttpData.Text(text, charset)   => UIO(Unpooled.copiedBuffer(text, charset))
+        case HttpData.BinaryChunk(data)     => UIO(Unpooled.copiedBuffer(data.toArray))
+        case HttpData.BinaryByteBuf(data)   => UIO(data)
+        case HttpData.Empty                 => UIO(Unpooled.EMPTY_BUFFER)
+        case HttpData.BinaryStream(stream)  =>
+          stream
+            .asInstanceOf[ZStream[Any, Throwable, ByteBuf]]
+            .fold(Unpooled.compositeBuffer())((c, b) => c.addComponent(b))
+        case HttpData.RandomAccessFile(raf) =>
+          effectBlocking {
+            val fis                      = new FileInputStream(raf().getFD)
+            val fileContent: Array[Byte] = new Array[Byte](raf().length().toInt)
+            fis.read(fileContent)
+            Unpooled.copiedBuffer(fileContent)
+          }
+      }
+  }
+
+  private[zhttp] final class UnsafeContent(private val httpContent: HttpContent) extends AnyVal {
+    def content: ByteBuf = httpContent.content()
+
+    def isLast: Boolean = httpContent.isInstanceOf[LastHttpContent]
+  }
+
+  private[zhttp] final class UnsafeChannel(private val ctx: ChannelHandlerContext) extends AnyVal {
+    def read(): Unit = ctx.read(): Unit
+  }
+
+  private[zhttp] final case class Incoming(unsafeRun: (UnsafeChannel => UnsafeContent => Unit) => Unit)
+      extends HttpData {
+    def encode: ZIO[Any, Nothing, ByteBuf] = for {
+      body <- ZIO.effectAsync[Any, Nothing, ByteBuf](cb =>
+        unsafeRun(ch => {
+          val buffer = Unpooled.compositeBuffer()
+          msg => {
+            buffer.writeBytes(msg.content)
```

Comment:
```suggestion
            buffer.addComponent(msg.content)
```

For better performance.

Reply:
Done